src/test/tools/pmemobjcli/pmemobjcli.c

```diff
@@ -1733,6 +1907,285 @@ pocli_pmemobj_tx_errno(struct pocli_ctx *ctx, struct pocli_args *args)
+	struct pobj_action *actp = NULL;
+	uint64_t type_num = 0;
+	uint64_t flags = 0;
+	size_t size = 0;
+	uint64_t class_id = 0;
+	int xalloc_zero = 0;
+	enum pocli_ret ret;
+
+	ret = pocli_args_obj(ctx, args, 1, &oidp);
+	if (ret)
+		return ret;
+
+	ret = pocli_args_act(ctx, args, 2, &actp);
+	if (ret)
+		return ret;
+
+	ret = pocli_args_number(args, 3, &type_num);
+	if (ret)
+		return ret;
+
+	ret = pocli_args_size(args, 4, &size);
+	if (ret)
+		return ret;
+
+	ret = pocli_args_class_id(args, 5, &class_id);
+	if (ret)
+		return ret;
+
+	ret = pocli_args_flag(args, 6, &xalloc_zero);
+	if (ret)
+		return ret;
+
+	if (xalloc_zero != 0)
+		flags = class_id + 1;
+	else
+		flags = class_id;
+
+	PMEMoid r = pmemobj_xreserve(ctx->pop, actp, size, type_num, flags);
+
+	if (oidp != NULL)
+	*oidp = r;
+
+	pocli_printf(ctx, "%s(%s, %s, %zu, %llu, %x): %d\n", args->argv[0],
+		args->argv[1], args->argv[2], size, type_num, flags, r);
+
+	return ret;
+}
+
+/*
+ * pocli_pmemobj_defer_free -- pmemobj_defer_free() command
+ */
+static enum pocli_ret
+pocli_pmemobj_defer_free(struct pocli_ctx *ctx, struct pocli_args *args)
+{
+	if (args->argc != 3)
+		return POCLI_ERR_ARGS;
+
+	PMEMoid *oidp = NULL;
+	struct pobj_action *actp = NULL;
+	enum pocli_ret ret;
+
+	ret = pocli_args_obj(ctx, args, 1, &oidp);
+	if (ret)
+		return ret;
+
+	ret = pocli_args_act(ctx, args, 2, &actp);
+	if (ret)
+		return ret;
+
+	if (oidp == NULL)
+		return pocli_err(ctx, POCLI_ERR_ARGS, "NULL OID not allowed\n");
+
+	pmemobj_defer_free(ctx->pop, *oidp, actp);
+
+	pocli_printf(ctx, "%s(%s, %s)\n", args->argv[0], args->argv[1],
+		args->argv[2]);
+
+	return ret;
+}
+
+/*
+ * pocli_pmemobj_set_value -- pmemobj_set_value() command
+ */
+static enum pocli_ret
+pocli_pmemobj_set_value(struct pocli_ctx *ctx, struct pocli_args *args)
+{
+	if (args->argc != 5)
+		return POCLI_ERR_ARGS;
+
+	PMEMoid *oidp = NULL;
+	struct pobj_action *actp = NULL;
+	uint64_t *ptr;
+	uint64_t value;
+	enum pocli_ret ret;
+
+	ret = pocli_args_obj(ctx, args, 1, &oidp);
+	if (ret)
+		return ret;
+
+	ret = pocli_args_act(ctx, args, 2, &actp);
+	if (ret)
+		return ret;
+	ret = pocli_args_number(args, 3, &value);
+	if (ret)
+		return ret;
+
+	ptr = (uint64_t *)pmemobj_direct(*oidp);
+
+	pmemobj_set_value(ctx->pop, actp, ptr, value);
+
+	pocli_printf(ctx, "%s(%s, %s, %llu, %llu)\n", args->argv[0],
+		args->argv[1], args->argv[2], ptr, value);
+
+	return ret;
+}
+
+/*
+ * pocli_pmemobj_publish -- pmemobj_publish() command
+ */
+static enum pocli_ret
+pocli_pmemobj_publish(struct pocli_ctx *ctx, struct pocli_args *args)
+{
+	if (args->argc != 4)
+		return POCLI_ERR_ARGS;
+
+	struct pobj_action *actp_pub = NULL;
+	size_t actcnt = 0;
+	enum pocli_ret ret;
+	unsigned g_pub[2];
+
+	ret = pocli_args_act_pub(ctx, args, 1, &actp_pub, g_pub);
+	if (ret)
+		return ret;
+
+	ret = pocli_args_act_pub(ctx, args, 2, &actp_pub, g_pub);
+	if (ret)
+		return ret;
+
+	ret = pocli_args_size(args, 3, &actcnt);
+	if (ret)
+		return ret;
+
```

Comment:
This implementation
1) doesn't match the usage string
2) is over-complicated and I'm not even sure is correct. Copy/pasting pocli_args_act and pocli_reserv_id into pocli_args_act_pub is ugly.

This op should accept only 2 args - the first action and number of following actions. It should also verify that all actions in the range are valid. It can be achieved by a) looking at "pobj_action.type" and b) zeroing actions after publish.